New isolatedDeclarations error with generic `this` parameter inference

## Steps to reproduce

```typescript
export function fromProviders<T>() {
  return {
    tearDown(this: {state: T}): void {}
  };
}
```

## Behavior with `typescript@6.0`

no error, .d.ts emitted is:

```typescript
export declare function fromProviders<T>(): {
    tearDown(this: {
        state: T;
    }): void;
};
```

[link](https://www.typescriptlang.org/play/?isolatedDeclarations=true&isolatedModules=true#code/KYDwDg9gTgLgBAMwK4DsDGMCWEWKhAWwAV8A3TAE2CgGcAeAFQD4AKASjgG8AoOOKYDCRRcPPnxjAAhlAAiEAO4oWMABaYaALi40YUydoYBfNttIRKXI7zhGA3NyNA)

## Behavior with `tsgo`

```
error TS9008: Method must have an explicit return type annotation with --isolatedDeclarations.
```

[link](https://tsgo.sxzz.dev/#eNpFjsGqwkAMRX9lyEqh+AHzHq7c68KddTFMMzpSJyVJq1D676ZKdRXCuXDOCAn86QQqkUrKl81NqEAFY12cqyHSvcst8r7TTEVq8O5NZpaF2qDY7DC2gcN3oNxjtYyaH1zYjKa6THCuzIuiGxUz4rMjVpf6Eue1S0z3A9OQG2T5P25X64+aUXsuS4Zi4B09ykqvWSxO1JK8O05r7wbKjRtNZb6/t/BcQQQPZhvszPWi9jzAp9AKTi+1s1rK)

--- a/internal/checker/pseudotypenodebuilder.go
+++ b/internal/checker/pseudotypenodebuilder.go
@@ -549,7 +549,12 @@ func (b *NodeBuilderImpl) pseudoTypeEquivalentToType(t *pseudochecker.PseudoType
 }
 
 func (b *NodeBuilderImpl) pseudoParametersEquivalentToParameters(params []*pseudochecker.PseudoParameter, targetSig *Signature, reportErrors bool, nonParamErrorLocation *ast.Node) bool {
-	if targetSig.thisParameter != nil && ast.IsThisIdentifier(params[0].Name) {
+	if targetSig.thisParameter != nil && len(params) == 0 {
+		if reportErrors {
+			b.ctx.tracker.ReportInferenceFallback(nonParamErrorLocation) // missing `this` param
+		}
+		return false;
+	} else if targetSig.thisParameter != nil && ast.IsThisIdentifier(params[0].Name) {
 		targetParam := targetSig.thisParameter
 		paramType := b.ch.getTypeOfParameter(targetParam)
 		if !b.pseudoTypeEquivalentToType(params[0].Type, paramType, params[0].Optional, false) {
@@ -569,10 +574,16 @@ func (b *NodeBuilderImpl) pseudoParametersEquivalentToParameters(params []*pseud
 		if reportErrors {
 			b.ctx.tracker.ReportInferenceFallback(nonParamErrorLocation)
 		}
-		return false
+		return false // TODO: spread tuple params may mess with this check
 	}
 	for i, p := range params {
 		targetParam := targetSig.parameters[i]
+		if p.Optional != b.ch.isOptionalParameter(targetParam.ValueDeclaration) {
+			if reportErrors {
+				b.ctx.tracker.ReportInferenceFallback(p.Name.Parent)
+			}
+			return false
+		}
 		paramType := b.ch.getTypeOfParameter(targetParam)
 		if !b.pseudoTypeEquivalentToType(p.Type, paramType, p.Optional, false) {
 			if reportErrors {

--- a/internal/checker/pseudotypenodebuilder.go
+++ b/internal/checker/pseudotypenodebuilder.go
@@ -450,21 +450,9 @@ func (b *NodeBuilderImpl) pseudoTypeEquivalentToType(t *pseudochecker.PseudoType
 					// Target property type doesn't have a single call signature; can't validate
 					continue
 				}
-				if len(targetSig.parameters) != len(d.Parameters) {
-					if reportErrors {
-						b.ctx.tracker.ReportInferenceFallback(e.Name.Parent)
-					}
+				paramEq := b.pseudoParametersEquivalentToParameters(d.Parameters, targetSig, reportErrors, e.Name.Parent)
+				if !paramEq {
 					return false
-				}
-				for i, p := range d.Parameters {
-					targetParam := targetSig.parameters[i]
-					paramType := b.ch.getTypeOfParameter(targetParam)
-					if !b.pseudoTypeEquivalentToType(p.Type, paramType, p.Optional, false) {
-						if reportErrors {
-							b.ctx.tracker.ReportInferenceFallback(e.Name.Parent)
-						}
-						return false
-					}
 				}
 				targetPredicate := b.ch.getTypePredicateOfSignature(targetSig)
 				if targetPredicate != nil {
@@ -533,27 +521,9 @@ func (b *NodeBuilderImpl) pseudoTypeEquivalentToType(t *pseudochecker.PseudoType
 			}
 			return false
 		}
-		if len(targetSig.parameters) != len(pt.Parameters) {
-			if reportErrors {
-				b.ctx.tracker.ReportInferenceFallback(pt.Signature)
-			}
-			return false // TODO: spread tuple params may mess with this check
-		}
-		for i, p := range pt.Parameters {
-			targetParam := targetSig.parameters[i]
-			if p.Optional != b.ch.isOptionalParameter(targetParam.ValueDeclaration) {
-				if reportErrors {
-					b.ctx.tracker.ReportInferenceFallback(p.Name.Parent)
-				}
-				return false
-			}
-			paramType := b.ch.getTypeOfParameter(targetParam)
-			if !b.pseudoTypeEquivalentToType(p.Type, paramType, p.Optional, false) {
-				if reportErrors {
-					b.ctx.tracker.ReportInferenceFallback(p.Name.Parent)
-				}
-				return false
-			}
+		paramEq := b.pseudoParametersEquivalentToParameters(pt.Parameters, targetSig, reportErrors, pt.Signature)
+		if !paramEq {
+			return false
 		}
 		targetPredicate := b.ch.getTypePredicateOfSignature(targetSig)
 		if targetPredicate != nil {
@@ -576,6 +546,42 @@ func (b *NodeBuilderImpl) pseudoTypeEquivalentToType(t *pseudochecker.PseudoType
 	default:
 		return false
 	}
+}
+
+func (b *NodeBuilderImpl) pseudoParametersEquivalentToParameters(params []*pseudochecker.PseudoParameter, targetSig *Signature, reportErrors bool, nonParamErrorLocation *ast.Node) bool {
+	if targetSig.thisParameter != nil && ast.IsThisIdentifier(params[0].Name) {
+		targetParam := targetSig.thisParameter
+		paramType := b.ch.getTypeOfParameter(targetParam)
+		if !b.pseudoTypeEquivalentToType(params[0].Type, paramType, params[0].Optional, false) {
+			if reportErrors {
+				b.ctx.tracker.ReportInferenceFallback(params[0].Name.Parent)
+			}
+			return false
+		}
+		params = params[1:]
+	} else if targetSig.thisParameter != nil {
+		if reportErrors {
+			b.ctx.tracker.ReportInferenceFallback(nonParamErrorLocation)
+		}
+		return false
+	}
+	if len(targetSig.parameters) != len(params) {
+		if reportErrors {
+			b.ctx.tracker.ReportInferenceFallback(nonParamErrorLocation)
+		}
+		return false
+	}
+	for i, p := range params {
+		targetParam := targetSig.parameters[i]
+		paramType := b.ch.getTypeOfParameter(targetParam)
+		if !b.pseudoTypeEquivalentToType(p.Type, paramType, p.Optional, false) {
+			if reportErrors {
+				b.ctx.tracker.ReportInferenceFallback(p.Name.Parent)
+			}
+			return false
+		}
+	}
+	return true
 }
 
 func isStructuralPseudoType(t *pseudochecker.PseudoType) bool {

--- a/internal/checker/pseudotypenodebuilder.go
+++ b/internal/checker/pseudotypenodebuilder.go
@@ -553,7 +553,7 @@ func (b *NodeBuilderImpl) pseudoParametersEquivalentToParameters(params []*pseud
 		if reportErrors {
 			b.ctx.tracker.ReportInferenceFallback(nonParamErrorLocation) // missing `this` param
 		}
-		return false;
+		return false
 	} else if targetSig.thisParameter != nil && ast.IsThisIdentifier(params[0].Name) {
 		targetParam := targetSig.thisParameter
 		paramType := b.ch.getTypeOfParameter(targetParam)

--- a/testdata/baselines/reference/compiler/isolatedDeclarationsThisParameterWithCapturedGeneric.js
+++ b/testdata/baselines/reference/compiler/isolatedDeclarationsThisParameterWithCapturedGeneric.js
@@ -1,0 +1,23 @@
+//// [tests/cases/compiler/isolatedDeclarationsThisParameterWithCapturedGeneric.ts] ////
+
+//// [isolatedDeclarationsThisParameterWithCapturedGeneric.ts]
+export function fromProviders<T>() {
+  return {
+    tearDown(this: {state: T}): void {}
+  };
+}
+
+//// [isolatedDeclarationsThisParameterWithCapturedGeneric.js]
+export function fromProviders() {
+    return {
+        tearDown() { }
+    };
+}
+
+
+//// [isolatedDeclarationsThisParameterWithCapturedGeneric.d.ts]
+export declare function fromProviders<T>(): {
+    tearDown(this: {
+        state: T;
+    }): void;
+};

--- a/testdata/baselines/reference/compiler/isolatedDeclarationsThisParameterWithCapturedGeneric.symbols
+++ b/testdata/baselines/reference/compiler/isolatedDeclarationsThisParameterWithCapturedGeneric.symbols
@@ -1,0 +1,16 @@
+//// [tests/cases/compiler/isolatedDeclarationsThisParameterWithCapturedGeneric.ts] ////
+
+=== isolatedDeclarationsThisParameterWithCapturedGeneric.ts ===
+export function fromProviders<T>() {
+>fromProviders : Symbol(fromProviders, Decl(isolatedDeclarationsThisParameterWithCapturedGeneric.ts, 0, 0))
+>T : Symbol(T, Decl(isolatedDeclarationsThisParameterWithCapturedGeneric.ts, 0, 30))
+
+  return {
+    tearDown(this: {state: T}): void {}
+>tearDown : Symbol(tearDown, Decl(isolatedDeclarationsThisParameterWithCapturedGeneric.ts, 1, 10))
+>this : Symbol(this, Decl(isolatedDeclarationsThisParameterWithCapturedGeneric.ts, 2, 13))
+>state : Symbol(state, Decl(isolatedDeclarationsThisParameterWithCapturedGeneric.ts, 2, 20))
+>T : Symbol(T, Decl(isolatedDeclarationsThisParameterWithCapturedGeneric.ts, 0, 30))
+
+  };
+}

--- a/testdata/baselines/reference/compiler/isolatedDeclarationsThisParameterWithCapturedGeneric.types
+++ b/testdata/baselines/reference/compiler/isolatedDeclarationsThisParameterWithCapturedGeneric.types
@@ -1,0 +1,16 @@
+//// [tests/cases/compiler/isolatedDeclarationsThisParameterWithCapturedGeneric.ts] ////
+
+=== isolatedDeclarationsThisParameterWithCapturedGeneric.ts ===
+export function fromProviders<T>() {
+>fromProviders : <T>() => { tearDown(this: { state: T; }): void; }
+
+  return {
+>{    tearDown(this: {state: T}): void {}  } : { tearDown(this: { state: T; }): void; }
+
+    tearDown(this: {state: T}): void {}
+>tearDown : (this: { state: T; }) => void
+>this : { state: T; }
+>state : T
+
+  };
+}

--- a/testdata/tests/cases/compiler/isolatedDeclarationsThisParameterWithCapturedGeneric.ts
+++ b/testdata/tests/cases/compiler/isolatedDeclarationsThisParameterWithCapturedGeneric.ts
@@ -1,0 +1,7 @@
+// @declaration: true
+// @isolatedDeclarations: true
+export function fromProviders<T>() {
+  return {
+    tearDown(this: {state: T}): void {}
+  };
+}

